torch/csrc/autograd/FunctionsManual.cpp

```diff
@@ -2445,43 +2445,50 @@ Tensor det_backward(const Tensor & grad, const Tensor& self, const Tensor& det)
   };
 
   auto nonsingular_case_backward = [&](const Tensor& grad, const Tensor& self, const Tensor& det) -> Tensor {
-    return unsqueeze_multiple(grad * det, {-1, -2}, self.dim()) * self.inverse().transpose(-2, -1);
+    // Derived from Jacobi's formula for partial derivatives, which can be
+    // found at https://en.wikipedia.org/wiki/Jacobi%27s_formula
+    auto det_grad = (unsqueeze_multiple(det, {-1, -2}, self.dim()) * self.inverse().transpose(-2, -1)).conj();
+    return unsqueeze_multiple(grad, {-1, -2}, self.dim()) * det_grad;
   };
 
   if (self.dim() == 2) {
-    if (det.item<double>() == 0) {
+    if (det.eq(0).item<bool>()) {
       return singular_case_backward(grad, self, det);
     } else {
       return nonsingular_case_backward(grad, self, det);
     }
   } else {
-    auto nonzero_det_indices = at::native::toListOfOptionalTensors(at::where(det));
```

Comment:
Note: The `nonzero` call within `where` does not have complex number support. So I decided to try using a bool mask to select singular/nonsingular parts of the inputs and outputs, instead of indices. I think that in principle the bool mask should be faster, for several reasons:

* The mask is easily invertible, so we can avoid having to recreate the zero mask from scratch. Previously, we were calling `where` twice to create both masks.
* Creating a boolean mask is more space efficient than a list of index tensors. It seems like it would also be faster ideally
* I *think* (but not sure) that calling `index` with a boolean mask could also be faster than calling it with a list of indices--potentially less data to iterate through.

One thing that could slow performance is the fact that we have to make an `all` call to see if the matrices are either all-singular or all-nonsingular.

I will measure performance before/after this PR and post the results. If it turns out that the original was faster, we can just wait for #50655 to add complex support for `nonzero`.